### DISCUSSION TITLE
Can mobx include toGenerator type?

### DISCUSSION THREAD

**Discussion by chrisands** [mobxjs/mobx#2804] [Ideas]
Can `mobx`  include `toGenerator` type from `mobx-state-tree`?

```ts
export function* toGenerator<R>(p: Promise<R>): Generator<Promise<R>, R, R> {
  return (yield p) as R
}
```

It helps to keep type after yield

```ts
const result = yield* toGenerator(api.getResult())
// result will be return type of api.getResult() instead of type `any`
```

**Comment by mweststrate** [comment]
See flowResult api

**Reply by kubk** [reply]
@mweststrate Documentation isn't complete: https://mobx.js.org/actions.html#using-flow-instead-of-async--await-

![image](https://user-images.githubusercontent.com/22447849/107939672-1ef4bd80-6f98-11eb-9b70-a3e17f3b2f67.png)

Am I correct that `fetchGithubProjectsSomehow()` should be wrapped with `flowResult` ?

**Reply by kubk** [reply]
@mweststrate I've created a simple reproduction, `flowResult` loses types: https://www.typescriptlang.org/play?#code/PTAEGEHsAcE9QGSmgQwM4BcCmoBmAnSAW1AFlIAjAD1A1mizQC4AoLK6SfDW+ncFADsAxlgA2YlBTFYACoSIBLNFgA8AFQB8oALyh5xZWq2JQAb1DChosQAoAlE1AA3SIoAmoAL5sOXHu5YwpL4OLgAriIYipCCeGKQAO4ASozhYhgamrahaOkYTuqOoOqg7NiC7migAOJYglj4KBhcqkKwADSgioK4jaDJXe3aAPwQ1uKS0nIKRqrJ2oVlVBVV4yKTUjIGSiptgrCjJaBODc6NANwsLCAljBgswrGYoADmWBgAcuFEFP16Dl02h2RgAdLlIGJzrYAIz2K4sCJRGJxABUKGE0VigLMLFA+Mszx4aAAFpB0u4AEJYb6-f6gWCKcSeXAJFJpDK2d5fH5-fAOeGgW4AZTJFNAf1Agl5-R6mCwKE8kFwoAA5O1VSwvEA

**Comment by mweststrate** [comment]
Ok, I think in that case flowResult can be overloaded? PR welcome. I see
that it doesn't do exactly the same as toGenerator, but I think the same
utility can take care of both cases, to avoid some headaches for user
whento use which?
